core/src/Streamly/Internal/Data/Scan.hs

```diff
@@ -0,0 +1,373 @@
+--
+-- >>> Stream.toList $ Stream.runScan (Scan.map (+1)) $ Stream.fromList [1..5::Int]
+-- [2,3,4,5,6]
+--
+{-# INLINE map #-}
+map :: Monad m => (a -> b) -> Scan m a b
+map f = mapM (return Prelude.. f)
+
+{- HLINT ignore "Redundant map" -}
+
+-- | An identity scan producing the same output as input.
+--
+-- >>> identity = Scan.map Prelude.id
+--
+-- >>> Stream.toList $ Stream.runScan (Scan.identity) $ Stream.fromList [1..5::Int]
+-- [1,2,3,4,5]
+--
+{-# INLINE identity #-}
+identity :: Monad m => Scan m a a
+identity = map Prelude.id
+
+instance Monad m => Category (Scan m) where
+    {-# INLINE id #-}
+    id = identity
+
+    {-# INLINE (.) #-}
+    (.) = compose
+
+-------------------------------------------------------------------------------
+-- Applicative Zip
+-------------------------------------------------------------------------------
+
+{-# ANN type TeeWith Fuse #-}
+data TeeWith sL sR = TeeWith !sL !sR
+
+-- XXX zipWith?
+
+-- | Connect two scans in parallel. Distribute the input across two scans and
+-- zip their outputs. If the scan filters the output, 'Nothing' is emitted
+-- otherwise 'Just' is emitted. The scan stops if any of the scans stop.
+--
+-- >>> Stream.toList $ Stream.runScan (Scan.teeWithMay (,) Scan.identity (Scan.map (\x -> x * x))) $ Stream.fromList [1..5::Int]
+-- [(Just 1,Just 1),(Just 2,Just 4),(Just 3,Just 9),(Just 4,Just 16),(Just 5,Just 25)]
+--
+{-# INLINE_NORMAL teeWithMay #-}
+teeWithMay :: Monad m =>
+    (Maybe b -> Maybe c -> d) -> Scan m a b -> Scan m a c -> Scan m a d
+teeWithMay f (Scan stepL initialL) (Scan stepR initialR) =
+    Scan step (TeeWith initialL initialR)
+
+    where
+
+    step (TeeWith sL sR) a = do
+        resL <- stepL sL a
+        resR <- stepR sR a
+        return
+            $ case resL of
+                  Yield sL1 bL ->
+                    case resR of
+                        Yield sR1 bR ->
+                            Yield (TeeWith sL1 sR1) (f (Just bL) (Just bR))
+                        Skip sR1 ->
+                            Yield (TeeWith sL1 sR1) (f (Just bL) Nothing)
+                        Stop -> Stop
+                  Skip sL1 ->
+                    case resR of
+                        Yield sR1 bR ->
+                            Yield (TeeWith sL1 sR1) (f Nothing (Just bR))
+                        Skip sR1 ->
+                            Yield (TeeWith sL1 sR1) (f Nothing Nothing)
+                        Stop -> Stop
+                  Stop -> Stop
+
+-- | Produces an output only when both the scans produce an output. If any of
+-- the scans skips the output then the composed scan also skips. Stops when any
+-- of the scans stop.
+--
+-- >>> Stream.toList $ Stream.runScan (Scan.teeWith (,) Scan.identity (Scan.map (\x -> x * x))) $ Stream.fromList [1..5::Int]
+-- [(1,1),(2,4),(3,9),(4,16),(5,25)]
+--
+{-# INLINE_NORMAL teeWith #-}
+teeWith :: Monad m =>
+    (b -> c -> d) -> Scan m a b -> Scan m a c -> Scan m a d
+teeWith f s1 s2 =
```

Comment:
Zipping based on outputs is a different combinator, not implemented. See the zipWith in pipe module which has this behavior, it is more complicated to implement and not sure where it would be useful, so not implemented yet.